manipulation/src/torch/csrc/jit/import.cpp

```diff
@@ -0,0 +1,228 @@
+    }
+    n.ts_(sym, std::move(tensors));
+    break;
+  }
+  case onnx::AttributeProto_AttributeType_GRAPH:
+    n.g_(sym, graphProtoToGraph(ap.g()));
+    break;
+  case onnx::AttributeProto_AttributeType_GRAPHS: {
+    std::vector<std::shared_ptr<Graph>> graphs;
+    for (int i = 0; i < ap.graphs_size(); i++) {
+      graphs.push_back(graphProtoToGraph(ap.graphs(i)));
+    }
+    n.gs_(sym, std::move(graphs));
+    break;
+  }
+  case onnx::AttributeProto_AttributeType_UNDEFINED:
+    abort();
+    break;
+  }
+}
+
+void convertAttributes(onnx::NodeProto & np, jit::Node & n) {
+  for (int i = 0; i < np.attribute_size(); i++) {
+    convertAttribute(np.attribute(i), n);
+  }
+}
+
+std::vector<jit::Dimension> tensorShapeProtoToDimensions(const onnx::TypeProto_TensorShapeProto & tsp) {
+  std::vector<jit::Dimension> dims;
+  for (int i = 0; i < tsp.dim_size(); i++) {
+    if (tsp.dim(i).has_dim_value()) {
+      dims.push_back(jit::Dimension(true, tsp.dim(i).dim_value(), ""));
+    } else {
+      dims.push_back(jit::Dimension(false, -1, tsp.dim(i).dim_param()));
+    }
+  }
+  return dims;
+}
+
+TypePtr TypeProtoToTypePtr(const onnx::TypeProto & tp) {
+  auto tt = tp.tensor_type();
+  auto sh = tensorShapeProtoToDimensions(tt.shape());
+  auto et = protoTypeToATenType(tt.elem_type());
+  return std::make_shared<TensorType>(TensorType(et, sh));
+}
+
+Graph* gg = nullptr;
+
+std::unique_ptr<Graph> graphProtoToGraph(const onnx::GraphProto& gp) {
+  std::unique_ptr<Graph> g(new Graph());
+
+  gg = g.get();
+
+  // Values are created (as in `new Value(..)`) by the Node that
+  // outputs them. Therefore we initialize the Nodes and Values in
+  // several stages.
+  //
+  // 1) add all input (to the graph) Values, owned by the sentinel Param node
+  // 2) add all Nodes and their output Values, but don't intialize inputs
+  // 3) initialize inputs of all Nodes
+  // 4) initialize inputs of the Return sentinel node
+  // 5) fill in type info for graph outputs, and register them as outputs
+  // 5) fill in type info for Values from the value_info list in the graph
+
+  // In ONNX proto land, Values are just strings. We are going to make
+  // objects out of them, and equal strings must be mapped to the same
+  // Value object.
+  std::unordered_map<std::string, Value*> value_by_name_of;
+
+  // We initialize Node inputs in a separate pass from the Nodes
+  // themselves. To do so, we need to have access to the names of the
+  // inputs.
+  std::unordered_map<Node*, std::vector<std::string>> inputs_by_node;
+
+  for (int i = 0; i < gp.input_size(); i++) {
+    auto vip = gp.input(i);
+    auto v = g->addInput();
+    v->setType(TypeProtoToTypePtr(vip.type()));
+    v->setUniqueName(vip.name());
+    value_by_name_of[vip.name()] = v;
+  }
+
+  for (int i = 0; i < gp.node_size(); i++) {
+    auto np = gp.node(i);
+    auto n = g->create(stringToSymbol(np.op_type()), /* num_outputs = */ np.output_size());
```

Comment:
I know the other code in pytorch doesn't follow this convention but I prefer `auto* n = ...` to make it explicit that this is a pointer without having to know the signature of the rvalue expression. @ezyang  thoughts?

Reply:
sure, I'll do it